ui/jobs/player.ts

```diff
@@ -0,0 +1,481 @@
+      if (job !== newJob)
+        this.off('job-detail', wrapper);
+    });
+  }
+
+  private processPlayerChangedEvent(
+    { detail: data }: OverlayEventResponses['onPlayerChangedEvent'],
+  ): void {
+    this.id = data.id;
+    this.name = data.name;
+
+    // always update stuffs when player changed their jobs
+    const prevJob = this.job;
+    if (prevJob !== data.job) {
+      this.job = data.job;
+      this.emit('job', data.job);
+    }
+
+    // update level
+    if (this.level !== data.level) {
+      const prevLevel = this.level;
+      this.level = data.level;
+      this.emit('level', data.level, prevLevel);
+    }
+
+    // update hp
+    if (
+      prevJob !== data.job ||
+      this.hp !== data.currentHP ||
+      this.maxHp !== data.maxHP ||
+      this.shield !== data.currentShield
+    ) {
+      const prevHp = this.hp;
+      const prevShield = this.shield;
+      this.hp = data.currentHP;
+      this.maxHp = data.maxHP;
+      this.shield = data.currentShield;
+      this.emit('hp', {
+        hp: data.currentHP,
+        maxHp: data.maxHP,
+        prevHp: prevHp,
+        shield: data.currentShield,
+        prevShield: prevShield,
+      });
+    }
+
+    // update mp
+    if (
+      prevJob !== data.job ||
+      this.mp !== data.currentMP ||
+      this.maxMp !== data.maxMP
+    ) {
+      const prevMp = this.mp;
+      this.mp = data.currentMP;
+      this.maxMp = data.maxMP;
+      this.emit('mp', {
+        mp: data.currentMP,
+        maxMp: data.maxMP,
+        prevMp: prevMp,
+      });
+    }
+
+    // update cp
+    if (
+      prevJob !== data.job ||
+      this.cp !== data.currentCP ||
+      this.maxCp !== data.maxCP
+    ) {
+      const prevCp = this.cp;
+      this.cp = data.currentCP;
+      this.maxCp = data.maxCP;
+      this.emit('cp', {
+        cp: data.currentCP,
+        maxCp: data.maxCP,
+        prevCp: prevCp,
+      });
+    }
+
+    // update gp
+    if (
+      prevJob !== data.job ||
+      this.gp !== data.currentGP ||
+      this.maxGp !== data.maxGP
+    ) {
+      const prevGp = this.gp;
+      this.gp = data.currentGP;
+      this.maxGp = data.maxGP;
+      this.emit('gp', {
+        gp: data.currentGP,
+        maxGp: data.maxGP,
+        prevGp: prevGp,
+      });
+    }
+
+    if (
+      this.pos.x !== data.pos.x ||
+      this.pos.y !== data.pos.y ||
+      this.pos.z !== data.pos.z ||
+      this.rotation !== data.rotation
+    ) {
+      this.pos = data.pos;
+      this.rotation = data.rotation;
+      this.emit('pos', data.pos, data.rotation);
+    }
+
+    // update job details if there are
+    if (data.jobDetail && !isEqual(this.jobDetail, data.jobDetail)) {
+      this.jobDetail = data.jobDetail;
+      this.emit('job-detail', data.job, data.jobDetail);
+    }
+
+    this.emit('player', this);
+  }
+
+  private processPlayerStatsLogLine(line: string[]): void {
+    const matches = normalizeLogLine(line, logDefinitions.PlayerStats.fields);
+    // const stat = Object
```

Comment:
Is this old code and the comment should be removed?

Reply:
Oops!
The code commented is the clear way to do the conversion (but a bit hard to understand), but then I should fight with typescript type checking as it lost types after the `filter` part, so I am not very sure which is better, comparing the "assigning-every-field-we-need" approach below.

Reply:
I think assigning-every-field is ok.  It's definitely not as satisfying, but it's hard to convince typescript that you've set everything without explicitly setting everything.

Reply:
Yeah, you said it.
Okay, I'll remove the commented code here.